### ISSUE TITLE
网络部分

### ISSUE THREAD

**Issue by crystalYY** [crystalYY/MyBlog#1]
# 目录
- [http协议各版本的区别](#http--------)
  * [http1.1的特点](#http11---)
  * [http2的特点](#http2---)
- [浏览器缓存机制](#-------)
  * [跨域](#--)
    + [同源](#--)
    + [实现跨域的方法](#-------)
      - [使用代理](#----)
      - [CORS 跨域访问](#cors-----)
      - [JSONP](#jsonp)
  * [cookie](#cookie)
    + [cookie的属性选项](#cookie-----)
    + [设置cookie](#--cookie)
    + [删除修改cookie](#----cookie)
  * [http缓存](#http--)
  * [localStorage, sessionStorage](#localstorage--sessionstorage)
  * [application cache(manifest)](#application-cache-manifest-)
  * [get和post的区别](#get-post---)
  * [响应头，请求头结构](#---------)
  * [http报文的结构](#http-----)
* * *
**参考资料：**
[九种浏览器端缓存机制知多少](http://jixianqianduan.com/frontend-javascript/2015/12/28/nine-browser-cache-methods.html)
[同源](https://segmentfault.com/a/1190000015597029)
[聊一聊cookie](https://segmentfault.com/a/1190000004556040)
[cookie/session的运行机制](https://harttle.land/2015/08/10/cookie-session.html)
[http协议](http://www.ruanyifeng.com/blog/2016/08/http.html)
### http协议各版本的区别
[http协议](http://www.ruanyifeng.com/blog/2016/08/http.html)
#### http1.1的特点
1. 支持持久连接，无需再设置connection: keep-alive
2. 管道机制，客户端可以同时发送多个请求，但是服务器还是会按照顺序返回。同时为了区分各个数据包，增加了Content-Length字段，声明本次回应的数据长度
3. 可以在请求头或响应头使用Transfer-Encoding字段，实现分块传输
4. 新增了put, delete等方法
#### http2的特点
1. 是一个完全的二进制协议，http1.1的头信息为ASCII编码，为文本，数据体也可以是文本。
2. 实现了双向实时的多工通信
3. 有了数据流的概念，数据包可以不按顺序发送，但必须标记数据包所属的数据流ID
4. 客户端和服务器都可以不断开TCP连接，通过发送信号帧来取消某个数据流
5. 头信息压缩后发送，并且服务器和客户端同时维护一张信息表，包含cookie，user agent等，每次请求可以只发送索引号，而不必发送全部字段。
6. 服务器可以主动推送浏览器很可能需要的资源，比如解析html源码后，可能会发现需要图片等静态资源，服务器可以主动发送。
##### [同源](https://segmentfault.com/a/1190000015597029)
浏览器在发送请求时只接受同域的服务器响应的资源。同域指的是协议，域名和端口都相同，如果三个条件中有一个不一致就会导致跨域。
跨域只存在于客户端页面之间，服务器端代码请求则不受同源策略的限制

1. cookie，localStorage和indexDB无法获取
2. DOM不能取得
3. AJAX请求不能发送
    * 同域资源可读写;
    * 跨域请求会直接被浏览器拦截.(chrome下跨域请求不会发起, 其他浏览器一般是可发送跨域请求, 但响应被浏览器拦截)     
4. iframe限制

    * 可以访问同域资源, 可读写;
    * 访问跨域页面时, 只读.

###### 使用代理
虽然ajax和iframe受同源策略限制, 但服务器端代码请求, 却不受此限制, 我们可以基于此去伪造一个同源请求, 实现跨域的访问.
实现思路:

* 请求同域下的web服务器;
* web服务器像代理一样去请求真正的第三方服务器;
* 代理拿到数据过后, 直接返回给客户端ajax.
这样, 我们便拿到了跨域数据.
![](https://raw.githubusercontent.com/crystalYY/github_pic/master/img/Image.png)

###### CORS 跨域访问
HTML5带来的新的跨域请求的方式。保护资源只被可信的访问源以正确的方式访问。通过检查目的站点的响应头域，进而判断是否允许当前站点访问。通常服务器可以使用下列响应头来通知浏览器
![](https://raw.githubusercontent.com/crystalYY/github_pic/master/img/Image%20%5B2%5D.png)

CORS的解决办法是在服务端Response的HTTP头域加入资源的访问权限信息. 如: A站只需要在response头中加一个字段就能让B站跨站访问.
`access-control-allow-origin:*`
其中* 表示通配, 所有的域都能访问此资源, 如果严谨一些只允许B站访问:
`access-control-allow-origin:<B-DOMAIN>`
这样B站就可以直接访问此资源, 不需要JSONP 也不需要iframe了.
CORS需要指定METHOD访问, 对于GET和POST请求, 至少要指定以下三种methods, 如下:
`Access-Control-Allow-Methods: POST, GET, OPTIONS`
如果是POST请求, 且提交的数据类型是json, 那么, CORS需要指定headers.
`Access-Control-Allow-Headers: Content-Type`
CORS默认是不带cookie的, 设置以下字段将允许浏览器发送cookie.
`Access-Control-Allow-Credentials: true`
并需要在浏览器也设置`credentials：“include”`
除此之外, 为了跨站发送cookie等验证信息, access-control-allow-origin 字段将不允许设置为*, 它需要明确指定与请求网页一致的域名.同时, 请求网页中需要做如下显式设置才能真正发送`cookie.xhr.withCredentials = true;`

缺点：
1. 不兼容ie10以下的浏览器
2. 对于非简单请求，会发送两次请求，第一次为预检请求（preflight，一个options请求）,例如在header中添加自定义字段或者是使用
首部字段 Access-Control-Request-Method 告知服务器，实际请求将使用 POST 方法都可以触发预检请求，来保护不支持cors的浏览器，服务器需要成功响应2xx并完成了options请求，才会接下来进行下一次请求。
优点：
1. 支持所有类型的http请求
2. 可以通过onerror事件监听错误，并且可以在控制台看到报错信息，利于排查

简单请求：
![](https://raw.githubusercontent.com/crystalYY/github_pic/master/img/1.png)
![](https://raw.githubusercontent.com/crystalYY/github_pic/master/img/1.png)

###### JSONP

script标签并不受同源策略约束, 基于script 标签可做 jsonp 形式的访问, 可以通过第三方服务器生成动态的js代码**来回调本地的js方法**，而**方法中的参数则由第三方服务器在后台获取，并以JSON的形式填充到JS方法当中.** 即 JSON with Padding. 具体如下:

`<script type="text/javascript" src="https://www.targetDomain.com/jsonp?callback=callbackName"></script>`
优点：
1. 兼容性好
2. 只发送一次请求
缺点：
1. 只能用于获取资源，只读，类似于get请求
2. 错误处理机制不完善（这个一知半解）

#### cookie
参考资料：[聊一聊cookie](https://segmentfault.com/a/1190000004556040)
[cookie/session的运行机制](https://harttle.land/2015/08/10/cookie-session.html)

**定义**
Cookie是由http服务器设置的，保存在浏览器中的小型文本文件，其内容为一系列键值对，存放的为需要每次请求都发送给服务器的数据，比如身份认证信息。

**基本通信过程**
首先服务器给验证通过的浏览器通过生成sessionId来设置cookie，并通过response header（响应头）中的setCookie字段设置cookie。浏览器接收到cookie后会将其存储在存储cookie的文件夹中。
然后就是
发送请求之前首先检查是否有cookie -> 添加到请求头 -> 服务器收到cookie跟sessionid对比一致的话就发送数据

**cookie的格式**
可以通过document.cookie来获取非HttpOnly类型的cookie。cookie为一个字符串，由key=value键值对构成，之间由一个分号和一个空格隔开。

##### cookie的属性选项
cookie的属性包括：name, expires, domain, path, secure, HttpOnly.
**1. expires,设置失效时间（GMT格式）**
`expires=Thu, 25 Feb 2016 04:18:00 GMT`
不设置expires时，cookie默认有效期为session，在当前tag页面关闭后就失效，但是如果用户自己设置了浏览器的打开行为为从上次结束的地方开始，那么expires为session的cookie也不会失效。

在http1.1中，expires由max-age代替。expires为一个时间点（cookie失效时间 = expires）。max-age为一个以秒为单位的时间段（cookie失效时间 = cookie创建时间+max-age）
max-age默认为-1：有效期为一个session。
可以取三种值

* 负数：session
* 0： 删除
* 正数：cookie创建时刻+max-age

**2. domain 和 path**
某cookie的 domain为“baidu.com”, path为“/ ”，若请求的URL(URL 可以是js/html/img/css资源请求，但不包括 XHR 请求)的域名是“baidu.com”或其子域如“api.baidu.com”、“dev.api.baidu.com”，且 URL 的路径是“/ ”或子路径“/home”、“/home/login”，则浏览器会将此 cookie 添加到该请求的 cookie 头部中。

所以domain和path2个选项共同决定了cookie何时被浏览器自动添加到请求头部中发送出去。如果没有设置这两个选项，则会使用默认值。domain的默认值为url的host，path默认值为设置该cookie的网页所在的目录。

**3. secure**
设置cookie只在确保安全的请求中才会发送出去，比如请求为https。cookie默认不带secure选项。
`document.cookie = "name=huang; secure";`

**4. httpOnly**
设置cookie能否通过js访问，默认cookie不带httpOnly选项，浏览器可以通过js访问。设置了httpOnly之后，浏览器就不能通过js访问（读取，修改，删除）这个cookie。

在客户端是不能通过js代码去设置一个httpOnly类型的cookie的，这种类型的cookie只能通过服务端来设置。

凡是httpOnly类型的cookie，其控制台cookie页面上 HTTP 一列都会打上√

##### 设置cookie
**1. 服务器端**
通过响应头的Set-cookie字段设置cookie，但是一个set-cookie字段只能设置一个cookie。
![](https://raw.githubusercontent.com/crystalYY/github_pic/master/img/3.png)

服务端可以设置cookie 的所有选项：expires、domain、path、secure、HttpOnly

**2. 客户端**
通过document.cookie。

客户端可以设置cookie 的下列选项：expires、domain、path、secure（有条件：只有在https协议的网页中，客户端设置secure类型的 cookie 才能成功），但无法设置HttpOnly选项。
设置多个cookie：
![](https://raw.githubusercontent.com/crystalYY/github_pic/master/img/4.png)

##### 删除修改cookie
**修改cookie**
重新赋值即可，但是要保证path，domain的选项值和旧的cookie一致，否则不会修改而是添加了一个新的cookie

**删除cookie**
也是重新赋值，将expires设置为一个过去的时间点或者将max-age设为0，浏览器就会自动删除cookie。同样的，path/domain选项要和旧的保持一致。

#### http缓存
定义：基于http协议的浏览器文件级缓存机制，即针对文件的重复请求情况下，浏览器可以根据协议头判断是从服务器请求文件还是从本地读取文件。

**机制流程：**
1. 强缓存：通过expires或者cache-control:max-age，如果未过期，那么就直接读取http缓存文件，不会再发http请求，这是强缓存。
2. 协商缓存： 判断etag或last-modified是否存在，如果存在就分别在请求头带上if-none-match和if-modified-since字段发送给服务器，服务器先验证etag,如果etag更改，那么就重新给浏览器发送更新后的数据，返回200，否则返回304；如果不存在etag那么就对比last-modified的时间戳是否一致，不一致就返回更新后的内容，返回200，否则返回304

**etag和last-modified的区别**
1. ETag HTTP响应头是资源的特定版本的**标识符**，可以有助于防止资源的同时更新相互覆盖。可以在每次文件内容发生变化时，根据文件内容的hash值生成etag，比last-modified更加精确，但是消耗也更大，每次都要进行读写操作

2. last-modified存储的为一个时间点，只能精确到秒，秒之内的内容比如毫秒级别的更新只有etag才能检测，但是它消耗少。

**cache-control**
1. Cache-control: max-age=3246733
优先级高于expires，表示的是cache过期的相对时间

2. Cache-control: no-cache
可以存储在本地，但只是在与原始服务器进行新鲜度再验证之前，缓存不能被提供给客户端使用。即不使用本地缓存，需要使用协商缓存，先与服务器确认返回的响应是否被更改，再根据响应码确定本地缓存是否能被客户端使用。
3. Cache-control: no-store
缓存不可存储到本地
4. Cache-control: public
可以被所有用户缓存（多用户共享），比如终端浏览器或者CDN等中间代理服务器
5. Cache-control: private
只能被终端浏览器缓存（并且是私有缓存），不允许中继服务器进行缓存
#### localStorage, sessionStorage
localStorage和sessionStorage都可以通过setItem，getItem的方法或者.运算符属性的方法来设置或获取。
**设置**
```JavaScript
localStorage.setItem("name","caibin");   
localStorage.name = "caibin";
```
**获取**
```JavaScript
localStorage.getItem("name");   
localStorage.name;
```
**删除**
```javascript
localStorage.removeItem("name");
```
1. localStorage的有效期为永久，只能主动删除，如果想设置一个过期时间点自动删除的话，可以通过多设置一个locaStorage的字段（比如expires）为创建locaStorage的时间点（判断expires字段是否存在，不存在即为undefined的时候创建），每次访问网页的时候就查询一下当前时间是否大于创建的时间点，如果大于就删除。

2. sessionStorage为一个会话，关闭当前标签页就会自动删除
3. （自己的理解） localStorage可以存储一些譬如历史搜索等的信息，sessionStorage可以用于暂存编辑器编辑的数据

#### application cache(manifest)

**定义**
html5中的application cache接口是将大部分图片资源，js,css等静态资源放在manifest文件配置中，当页面打开时通过manifest文件来读取本地文件或是请求服务器文件来更新本地静态资源。

**优势**
applicationCache接口解决了由离线带来的部分难题，前提是需要访问的web页面至少被在线访问过一次。
* 离线浏览：用户可在离线时浏览完整网站
* 速度：缓存资源为本地资源，因此加载速度较快
* 服务器负载更少：只有服务器资源发生更改后，浏览器才会下载资源

**实现过程**
在线的情况下，浏览器发现html头部标签内有manifest属性，他会请求manifest文件。
![](https://raw.githubusercontent.com/crystalYY/github_pic/master/img/5.png)

![](https://raw.githubusercontent.com/crystalYY/github_pic/master/img/11.png)

如果是第一次访问APP，那么浏览器就会根据manifest文件的内容下载相应的资源并进行离线存储。
如果已经访问过APP并且资源已经离线存储了，那么浏览器就会使用离线的资源加载页面，然后浏览器会对比新的manifest文件与旧的manifest文件，如果文件没有发生改变，就不做任何操作，如果文件改变了，就会重新下载文件中的资源并进行离线存储，并在下一次失效。离线的情况下，浏览器就直接使用离线存储的资源。
![](https://raw.githubusercontent.com/crystalYY/github_pic/master/img/6.png)

#### get和post的区别
1. get是幂等安全的，post不是幂等的，连续调用会带来额外的影响
2. get通过url发送请求，长度限制比较大，post通过body发送，限制较小
3. get一般用于获取服务器数据，post用于将数据发送给服务器
4. get可以缓存，post在设置了过期时间expires等头字段时才能被缓存

#### 响应头，请求头结构

**请求头**
1. Origin: 指示了请求来自哪个站点。该字段仅指示服务器名称，不包含路径信息，用于CORS请求或POST请求，会包含协议
2. host: 指明服务器域名及端口号
![](https://raw.githubusercontent.com/crystalYY/github_pic/master/img/7.png)

3. If-Match：条件式请求头，在请求方法为get或head的情况下，服务器仅在请求的资源满足列出的ETag之一时才返回资源，为强比较算法，每一个比特都要相同。`
If-Match: <etag_value>, <etag_value>, …`，避免“空中碰撞”,如果哈希值不匹配，则意味着文档已经被编辑，抛出412前提条件失败错误。
4. If-None-Match: 条件式请求头，可以返回200或304，采用弱比较算法，即比特或内容相同都认为相同，比If-Modified-Since优先级高
5. Accept：请求头是客户端用来告知服务器自己可以处理的内容类型。服务器使用`Content-Type`响应头来通知客户端它的选择。
```
Accept: text/html

Accept: image/*

Accept: text/html, application/xhtml+xml, application/xml;q=0.9, */*;q=0.8
```
6. cookie：请求首部，包含先前由服务器通过`Set-Cookie`首部投放并存储到客户端的http cookie（如果用户禁用cookie，则会被完全移除）
7. Accept-Language 请求头允许客户端声明它可以理解的自然语言，以及优先选择的区域方言。
8. Accept-Charset 请求头用来告知（服务器）客户端可以处理的字符集类型。
9. HTTP 请求头 Accept-Encoding 会将客户端能够理解的内容编码方式——通常是某种压缩算法——进行通知。通过内容协商的方式，服务端会选择一个客户端提议的方式，使用并在
**响应报文首部 Content-Encoding**中通知客户端该选择。

**响应头**
1. ETag：http响应头，资源的特定版本标识符
2. Access-Control-Allow-Origin: 指定了该响应的资源是否被允许与给定的origin共享
3. Access-Control-Allow-Methods：在对预检请求的应答中明确了客户单所要访问的资源允许的方法或方法列表
4.  Access-Control-Allow-Headers：用于预检请求中，列出了将会在正式请求的Access-Control-Request-Headers字段中出现的首部信息
5. Content-Length: 是一个实体消息首部，用来指明发送给接收方的消息主体的大小，即用十进制数字表示的八位元组的数目。
6. Content-Type: 告诉客户端实际返回的内容的内容类型
`Content-Type: text/plain `
`Content-Type: text/html; charset=utf-8`
7. Content-Length 是一个实体消息首部，用来指明发送给接收方的消息主体的大小，即用十进制数字表示的八位元组的数目。
8. Content-Language 是一个 entity header （实体消息首部），用来说明访问者希望采用的语言或语言组合，这样的话用户就可以根据自己偏好的语言来定制不同的内容。`
<html lang="de">`
9. Set-Cookie
10. Last-Modified ,包含源头服务器认定的资源做出修改的日期及时间

#### http报文的结构
![](https://raw.githubusercontent.com/crystalYY/github_pic/master/img/8.png)

**1. 起始行**
报文的第一行就是起始行，在请求报文中用来说明要做些什么，在响应报文中说明出现了什么情况。

**2.首部字段**
起始行后面有零个或多个首部字段。每个首部字段都包含一个名字和一个值，为了便于解析，两者之间用冒号（:）来分隔。首部以一个空行结束。

**3.主体**
空行之后就是可选的报文主体了，其中包含了所有类型的数据。请求主体中包括了要发送给Web 服务器的数据；响应主体中装载了要返回给客户端的数据。起始行和首部都是文本形式且都是结构化的，而主体则不同，主体中可以包含任意的二进制数据（比如图片、视频、音轨、软件程序）。当然，主体中也可以包含文本。
